src/qat/ir/instructions.py

```diff
@@ -0,0 +1,1032 @@
+import numpy as np
+from compiler_config.config import InlineResultsProcessing
+from pydantic import (
+    BaseModel,
+    ConfigDict,
+    Field,
+    ValidationInfo,
+    field_serializer,
+    field_validator,
+)
+from typing_extensions import Annotated
+
+from qat.purr.compiler.devices import PulseChannel, PulseShapeType, QuantumComponent, Qubit
+
+# The following things from legacy instructions are unchanged, so just import for now.
+from qat.purr.compiler.instructions import AcquireMode, PostProcessType, ProcessAxis
+from qat.purr.utils.logger import get_default_logger
+
+
+### Sweeps
+class SweepOperation(BaseModel):
+    """Common parent for all things that need differentiating during a sweep."""
+
+    pass
+
+
+class SweepValue(SweepOperation):
+    name: str
+    value: Any
+
+
+### Variables
+def build_generated_name(existing_names=None, prefix=None):
+    if existing_names is None:
+        existing_names = set()
+
+    if prefix is None:
+        prefix = ""
+
+    if any(prefix) and not prefix.endswith("_"):
+        prefix = f"{prefix}_"
+
+    variable_name = f"{prefix}generated_name_{np.random.randint(np.iinfo(np.int32).max)}"
+    while variable_name in existing_names:
+        variable_name = (
+            f"{prefix}generated_name_{np.random.randint(np.iinfo(np.int32).max)}"
+        )
+
+    existing_names.add(variable_name)
+    return variable_name
+
+
+class Variable(BaseModel):
+    """
+    States that this value is actually a variable that should be fetched instead.
+    """
+
+    name: str
+    # TODO: we might consider limiting what types a variable can take...
+    var_type: Optional[type] = None
+    value: Any = None
+    model_config = ConfigDict(validate_assignment=True)
+
+    def __init__(self, name, var_type=None, value=None, **kwargs):
+        super().__init__(name=name, var_type=var_type, value=value)
+
+    @staticmethod
+    def with_random_name(existing_names=None, var_type=None, value=None):
+        return Variable(Variable.generate_name(existing_names), var_type, value)
+
+    @staticmethod
+    def generate_name(existing_names=None):
+        return build_generated_name(existing_names)
+
+    def __repr__(self):
+        return self.name
+
+    @field_serializer("var_type", when_used="json")
+    def _serialize_type(self, var_type: type) -> str:
+        # Types can't be (de)serialized, so we serialize as a string.
+        return var_type.__name__
+
+    @field_validator("var_type", mode="before")
+    @classmethod
+    def _validate_var_type(cls, var_type):
+        # Types can't be (de)serialized, so we use a validator to manually
+        # find the correct type.
+        if isinstance(var_type, str):
+            var_type = locate(var_type)
+        return var_type
+
+    @field_validator("value", mode="after")
+    @classmethod
+    def _validate_value_type(cls, value, val_info):
+        var_type = val_info.data["var_type"]
+        if var_type != None and not isinstance(value, var_type):
```

Comment:
How does this work if `value=None`? I.e. if I declare a `Variable("freq", var_type=float)` To be a frequency variable who's value should strictly be a float when set, but without setting it initially, e.g. to be used in a `Sweep`. :thinking:

Reply:
Good point, added support for this!